### PR TITLE
configs/*.json based configurations

### PR DESCRIPTION
**Pull request by jankii01**
Per Marcelo Salazar's proposal add configuration examples to a configs-folder.
By default the repo does not anymore have an mbed_app.json at all, but one should
copy one of the example configs, potentially modify it and use it.

Jenkinsfile was modified to do the copy operation at the beginning.

Add the emac for UBLOX to Wifi configs.

Reference:
    https://github.com/ARMmbed/mbed-os-example-client/issues/137

### CODE REVIEW

**Comment by jankii01** [comment]
Quite doesn't work yet, as the mbed_app.json isn't there in all cases now.

+ cp mbed-client-ci-scripts/test_automation/security.h .
+ rm -rf mbed-client-ci-scripts
+ mbed ls -a
mbed-os-example-client (git@github.com:ARMmbed/mbed-os-example-client.git#da6c44002fae)
|- atmel-rf-driver (ssh://git@github.com/armmbed/atmel-rf-driver.git#62c0c49d7a94)
|- esp8266-driver (ssh://git@github.com/ARMmbed/esp8266-driver.git#86ed47c8bf40)
|- mbed-client (ssh://git@github.com/ARMmbed/mbed-client.git#92bdf6a69958)
|  |- mbed-client-classic (ssh://git@github.com/ARMmbed/mbed-client-classic.git#42f18d977122)
|  `- mbed-client-mbed-tls (ssh://git@github.com/ARMmbed/mbed-client-mbed-tls.git#7e1b6d815038)
|- mbed-os (ssh://git@github.com/ARMmbed/mbed-os.git#d5de476f74dd)
|- mcr20a-rf-driver (ssh://git@github.com/ARMmbed/mcr20a-rf-driver.git#d8810e105d7d)
`- pal (ssh://git@github.com/ARMmbed/pal.git#4e46c0ea8706)
+ [ morpheus-wifi = morpheus ]
+ [ morpheus-wifi = morpheus-armcc ]
+ [ morpheus-wifi = morpheus-iar ]
+ [ morpheus-wifi = morpheus-wifi ]
+ sed -i -e s/ETHERNET/WIFI/g mbed_app.json
sed: can't read mbed_app.json: No such file or directory

@teetak or @kimlep need some help!

**Comment by jankii01** [comment]
Pulled back in the mbed_app.json to root, that sorted it. But we'd have to consolidate into something else in the future, I think. But, seems we have loads of different kinds of solutions now. Some are in Jenkins directly, some are in Jenkinsfile etc.

File: configs/eth_v4.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    "config": {
+        "network-interface":{
+            "help": "options are ETHERNET,WIFI,MESH_LOWPAN_ND,MESH_THREAD",
+            "value": "ETHERNET"
+        }
+    },
+    "macros": ["MBEDTLS_USER_CONFIG_FILE=\"mbedtls_mbed_client_config.h\""],
+    "target_overrides": {
+        "*": {
+            "target.features_add": ["LWIP", "NANOSTACK", "COMMON_PAL"],
```
Comment (teetak01) [inline configs/eth_v4.json]: should not need nanostack

File: configs/eth_v6.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "config": {
+        "network-interface":{
+            "help": "options are ETHERNET,WIFI,MESH_LOWPAN_ND,MESH_THREAD",
+            "value": "ETHERNET"
+        }
+    },
+    "macros": ["MBEDTLS_USER_CONFIG_FILE=\"mbedtls_mbed_client_config.h\""],
+    "target_overrides": {
+        "*": {
+            "target.features_add": ["LWIP", "NANOSTACK", "COMMON_PAL"],
```
Comment (teetak01) [inline configs/eth_v6.json]: should not need nanostack

File: configs/mesh_6lowpan.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        },
+        "mesh_radio_type": {
+        	"help": "options are ATMEL, MCR20",
+        	"value": "ATMEL"
+        }
+    },
+    "macros": ["MBEDTLS_USER_CONFIG_FILE=\"mbedtls_mbed_client_config.h\""],
+    "target_overrides": {
+        "*": {
+            "target.features_add": ["LWIP", "NANOSTACK", "LOWPAN_ROUTER", "COMMON_PAL"],
```
Comment (teetak01) [inline configs/mesh_6lowpan.json]: LWIP not needed here

File: configs/mesh_thread.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        },
+        "mesh_radio_type": {
+        	"help": "options are ATMEL, MCR20",
+        	"value": "ATMEL"
+        }
+    },
+    "macros": ["MBEDTLS_USER_CONFIG_FILE=\"mbedtls_mbed_client_config.h\""],
+    "target_overrides": {
+        "*": {
+            "target.features_add": ["LWIP", "NANOSTACK", "THREAD_ROUTER", "COMMON_PAL"],
```
Comment (teetak01) [inline configs/mesh_thread.json]: LWIP not needed here

File: configs/mesh_thread.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+            "mbed-mesh-api.thread-config-channel": 18,
+            "mbed-mesh-api.thread-config-panid": "0xBAAB",
+            "lwip.ipv4-enabled": false,
+            "lwip.ipv6-enabled": true,
```
Comment (teetak01) [inline configs/mesh_thread.json]: lwip specifications not needed

File: configs/mbed_app.json [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        "NUCLEO_F401RE": {
+            "wifi-tx": "PA_11",
+            "wifi-rx": "PA_12"
+        },
+        "NUCLEO_F411RE": {
+            "wifi-tx": "PA_11",
+            "wifi-rx": "PA_12"
+        }
+    }
+}
```
Comment (teetak01) [inline configs/mbed_app.json]: Is this file needed for something? It should be deleted INMO.

Instead the root folder should have minimal configuration for IPV4 LWIP as baseline. (So that the application will always build for something).

File: configs/wifi_v6.json [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        },
+        "wifi-rx": {
+            "help": "RX pin for serial connection to external device",
+            "value": "D0"
+        }
+    },
+    "macros": ["MBEDTLS_USER_CONFIG_FILE=\"mbedtls_mbed_client_config.h\""],
+    "target_overrides": {
+        "*": {
+            "target.features_add": ["NANOSTACK", "LOWPAN_ROUTER", "COMMON_PAL"],
```
Comment (teetak01) [inline configs/wifi_v6.json]: nanostack not needed

File: Jenkinsfile [hunk 8]
```diff
@@ -72,6 +72,9 @@ def buildStep(target, compilerLabel, toolchain, configName, connectiontype) {
         deleteDir()
         dir("mbed-os-example-client") {
           checkout scm
+          // Make us "backwards compatible" the easy way...
+          execute ("cp configs/example.json mbed_app.json")
```
Comment (teetak01) [inline Jenkinsfile]: not really needed. We instead need to have basic mbed_app.json in main folder.

**Comment by kimlep01** [comment]
I am seeing SOCKET_MODE : TCP again with 6lowpan ( same as with #139 ).

File: configs/eth_v4.json (hunk 1 above)
Comment (jankii01) [inline configs/eth_v4.json]: Unfortunately, it does! If one does NOT include Nanostack if fails:

jankii01@ubuntu:~/mbed/mbed-os-example-client$ ./build_all.sh 
Building project mbed-os-example-client (K64F, GCC_ARM)
Scan: .
Scan: FEATURE_LWIP
Scan: FEATURE_UVISOR
Scan: FEATURE_COMMON_PAL
Scan: FEATURE_BLE
Scan: FEATURE_STORAGE
Scan: FEATURE_THREAD_BORDER_ROUTER
Scan: FEATURE_THREAD_ROUTER
Scan: FEATURE_LOWPAN_BORDER_ROUTER
Scan: FEATURE_LOWPAN_ROUTER
Scan: FEATURE_LOWPAN_HOST
Scan: FEATURE_NANOSTACK_FULL
Scan: FEATURE_NANOSTACK
Scan: FEATURE_THREAD_END_DEVICE
Scan: mbed
Scan: env
Compile [  0.3%]: NanostackRfPhyAtmel.cpp
[ERROR] ./atmel-rf-driver/source/NanostackRfPhyAtmel.cpp:18:44: fatal error: nanostack/platform/arm_hal_phy.h: No such file or directory
compilation terminated.

-> it only works w/o NANOSTACK - IF YOU MOVE OUT THE other extra repos. This is due to the "compile all" nature of the mbed! If you move the extra RF-drivers to TARGET_IGNORE -folder, then it works. But that's another pain.

File: configs/wifi_v6.json [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        "NUCLEO_F401RE": {
+            "wifi-tx": "PA_11",
+            "wifi-rx": "PA_12"
+        },
+        "NUCLEO_F411RE": {
+            "wifi-tx": "PA_11",
+            "wifi-rx": "PA_12"
+        },
+        "UBLOX_EVK_ODIN_W2": {
+            "target.device_has": ["EMAC"]
```
Comment (0xc0170) [inline configs/wifi_v6.json]: EMAC is defined for ODIN by default

Comment (jankii01) [inline configs/wifi_v6.json]: Copied that from the mbed-os-example-wifi, that's on also there. 
To be removed from both, then?

File: configs/wifi_v4.json [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        "NUCLEO_F401RE": {
+            "wifi-tx": "PA_11",
+            "wifi-rx": "PA_12"
+        },
+        "NUCLEO_F411RE": {
+            "wifi-tx": "PA_11",
+            "wifi-rx": "PA_12"
+        },
+        "UBLOX_EVK_ODIN_W2": {
+            "target.device_has": ["EMAC"]
```
Comment (0xc0170) [inline configs/wifi_v4.json]: same here, EMAC defined by a target

File: configs/wifi_v6.json (hunk 9 above)
Comment (jankii01) [inline configs/wifi_v6.json]: Bigger issue seems to be our WiFi stacks don't support IPv6 at all?

File: configs/wifi_v4.json (hunk 10 above)
Comment (jankii01) [inline configs/wifi_v4.json]: Will remove it. And the whole wifi_v6.json.

File: configs/wifi_v6.json (hunk 9 above)
Comment (0xc0170) [inline configs/wifi_v6.json]: I'll fix wifi example , here's a reference to targets.json file, you cna see EMAC  there defined (wifi ON by default). https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L1310

Comment (0xc0170) [inline configs/wifi_v6.json]: > Bigger issue seems to be our WiFi stacks don't support IPv6 at all?

has not been tested neither verified (some bits are missing there) thus for now, unsupported
